compiler/AST/interfaces.cpp

```diff
@@ -0,0 +1,469 @@
+  bool wasConstrainedGeneric = parentFn->isConstrainedGeneric();
+  // follows introduceConstrainedTypes()
+  SET_LINENO(se);
+
+  // introduce a ConstrainedType
+  TypeSymbol* CT = ConstrainedType::build(astr("t_", isym->name));
+  parentFn->constrainedTypes.insertAtTail(new DefExpr(CT));
+
+  // add an interface constraint
+  ImplementsExpr* ie = new ImplementsExpr(new SymExpr(isym));
+  ie->implActuals.insertAtTail(new SymExpr(CT));
+  parentFn->interfaceConstraints.insertAtTail(ie);
+
+  // catch up with moveGlobalDeclarationsToModuleScope()
+  if (!wasConstrainedGeneric)
+    if (ModuleSymbol* mod = toModuleSymbol(parentFn->defPoint->parentSymbol))
+      {
+        FnSymbol* initFn = mod->initFn;
+        // follows insertProxyRef()
+        // wass todo this will not retain the exact location,
+        // wass if needed, do desugarInterfaceAsType() earlier in normalize
+        initFn->insertBeforeEpilogue(new CallExpr(PRIM_PROXY_REF, parentFn));
+        initFn->insertBeforeEpilogue(new CallExpr(PRIM_END_OF_STATEMENT));
+      }
+
+  return CT->type;
+}
+
+
+/////////////////////////////////////////////////////////////////////////////
+// Wrapper Function for ImplementsStmt
+//
+// An ImplementsStmt wrapper function helps us find an ImplementsStmt
+// for a ImplementsExpr interface constraint.
+//
+
+const char* implementsStmtWrapperName(InterfaceSymbol* isym) {
+  return astr("|", isym->name);
+}
+
+const char* interfaceNameForWrapperFn(FnSymbol* fn) {
+  return astr(fn->name + 1);
+}
+
+ImplementsStmt* implementsStmtForWrapperFn(FnSymbol* wrapFn) {
+  INT_ASSERT(wrapFn->hasFlag(FLAG_IMPLEMENTS_WRAPPER));
+  AList& body = wrapFn->body->body;
+  if (ImplementsStmt* istm = toImplementsStmt(body.head))
+    return istm;
+  else
+    INT_FATAL(wrapFn, "invalid implements wrapper function");
+
+  return NULL; //dummy
+}
+
+FnSymbol* wrapperFnForImplementsStmt(ImplementsStmt* istm) {
+  return toFnSymbol(istm->parentSymbol);
+}
+
+// Verify that the above functions work correctly.
+static void verifyWrapImplementsStmt(ImplementsStmt* istm,
+                                     InterfaceSymbol* isym,
+                                     FnSymbol* wrapFn) {
+  INT_ASSERT(wrapFn->name == implementsStmtWrapperName(isym));
+  INT_ASSERT(interfaceNameForWrapperFn(wrapFn) == isym->name);
+  INT_ASSERT(implementsStmtForWrapperFn(wrapFn) == istm);
+  INT_ASSERT(wrapperFnForImplementsStmt(istm) == wrapFn);
+}
+
+// To handle the cases where 'implExpr' requires resolving,
+// change wrapImplementsStatements as follows.
+// During normalization, set the type of the wrapper function's formal
+// ('arg' below) to dtAny. Detect this case when resolving a call
+// to a wrapper function, see implementsWrapperMatches() //wass correct?
+// There, replace the formal's type with the type of the actual
+// of the (by now hopefully resolved) implements statement.
+static Symbol* computeImplSym(Expr* implExpr) {
```

Comment:
what's an implSym? Is this trying to compute the symbol for the interface implemented? Could we rename it to better indicate what it is doing?